### PR TITLE
Fix truss under CheriABI

### PR DESCRIPTION
**Pull request by bsdjhb**
This fixes both mips64 and purecap truss to trace both mips64 and purecap processes.

### CODE REVIEW

File: sys/compat/freebsd64/freebsd64_process.c [hunk 1]
```diff
@@ -99,8 +99,8 @@ freebsd64_ptrace(struct thread *td, struct freebsd64_ptrace_args *uap)
 		struct dbreg dbreg;
 		struct fpreg fpreg;
 		struct reg reg;
-		char args[sizeof(td->td_sa.args)];
-		struct ptrace_sc_ret *psr;
```
Comment (bsdjhb) [inline sys/compat/freebsd64/freebsd64_process.c]: The '*psr' -> 'psr' bit is a mis-merge that could be a separate commit if really desired.

File: usr.bin/truss/Makefile [hunk 2]
```diff
@@ -9,4 +9,6 @@ LIBADD=	sysdecode
 #CFLAGS+= -I${.CURDIR} -I. -I${SRCTOP}/sys
 CFLAGS+= -I${SRCTOP}/sys
 
+_CHERI_CFLAGS+= -cheri-uintcap=addr
```
Comment (bsdjhb) [inline usr.bin/truss/Makefile]: I'm not sure if this is the right way to do this?

**Comment by bsdjhb** [comment]
We should perhaps be exporting bitmasks of tags for PT_GET_SC_ARGS and PT_GET_SC_RET.  For now I will punt on that though.

File: sys/kern/sys_process.c [hunk 3]
```diff
@@ -582,6 +583,19 @@ ptrace_sc_ret_to32(const struct ptrace_sc_ret *psr,
 }
 #endif /* COMPAT_FREEBSD32 */
 
+#if __has_feature(capabilities)
+static void
+ptrace_sc_ret_to64(const struct ptrace_sc_ret * __capability psr,
+    struct ptrace_sc_ret64 * __capability psr64)
+{
+
+	bzero((__cheri_fromcap void *)psr64, sizeof(*psr64));
```
Comment (arichardson) [inline sys/kern/sys_process.c]: I think this will generate a CToPtr relative to $ddc which is not ideal. Would be nice to have a capability bzero.

File: usr.bin/truss/Makefile (hunk 2 above)
Comment (arichardson) [reply usr.bin/truss/Makefile]: Probably fine for now. I guess we could use a CHERI_UINTCAP_MODE variable and change bsd.cheri.mk to pass `-cheri-uintcap=${CHERI_UINTCAP_MODE:Uoffset}` (not sure about the make syntax for a default value though).

File: usr.bin/truss/syscalls.c [hunk 4]
```diff
@@ -1514,7 +1521,7 @@ print_cmsgs(FILE *fp, pid_t pid, bool receive, struct msghdr *msghdr)
 	}
 	cmsgbuf = calloc(1, len);
 	if (get_struct(pid, (uintptr_t)msghdr->msg_control, cmsgbuf, len) == -1) {
-		fprintf(fp, "%p", msghdr->msg_control);
+		print_pointer(fp, (uintptr_t)msghdr->msg_control);
```
Comment (arichardson) [inline usr.bin/truss/syscalls.c]: @brooksdavis does `PRIxPTR` do the right thing for uintptr_t in purecap or does it print the offset?

File: usr.bin/truss/Makefile (hunk 2 above)
Comment (bsdjhb) [reply usr.bin/truss/Makefile]: I added CHERI_UINTCAP_MODE.

File: sys/kern/sys_process.c (hunk 3 above)
Comment (bsdjhb) [reply sys/kern/sys_process.c]: I agree that a bzero_c would be nice.  This pattern is used in several other places in this file currently so I was just reusing what was already there.

File: usr.bin/truss/syscalls.c (hunk 4 above)
Comment (bsdjhb) [reply usr.bin/truss/syscalls.c]: You changed %P to always print the address in 1322275559ae457e08b299290f4b3744e7394343.  Looking at plain '%p', it seems to have always printed the address and not the offset.

File: sys/compat/cheriabi/cheriabi_misc.c [hunk 5]
```diff
@@ -1897,9 +1895,11 @@ cheriabi_ptrace(struct thread *td, struct cheriabi_ptrace_args *uap)
 	case PT_VM_ENTRY:
 		error = COPYOUT(&r.pve, uap->addr, sizeof r.pve);
 		break;
+#endif
 	case PT_IO:
-		error = COPYOUT(&r.piod, uap->addr, sizeof r.piod);
+		error = copyoutcap(&r.piod, uap->addr, sizeof(r.piod));
```
Comment (brooksdavis) [inline sys/compat/cheriabi/cheriabi_misc.c]: In cases like this I prefer to update the `piod_len` field directly (at least I assume that's what can change) rather than copying the whole structure out.  My goal is to minimize the number of `copyoutcap` calls since each one is a potential capability leak that requires ongoing auditing.

File: sys/kern/sys_process.c (hunk 3 above)
Comment (brooksdavis) [reply sys/kern/sys_process.c]: The need for this reflects on the poor design of `kern_ptrace` where `addr` is overloaded and may be a user pointer or a kernel pointer.  I think this is fine for now, but we're overdue for an audit of `__cheri_fromcap` usage.

File: sys/compat/cheriabi/cheriabi_misc.c (hunk 5 above)
Comment (bsdjhb) [reply sys/compat/cheriabi/cheriabi_misc.c]: Ok.

File: sys/kern/sys_process.c (hunk 3 above)
Comment (bsdjhb) [reply sys/kern/sys_process.c]: In this case though it's more that we just don't have a bzero_c as the kernel capability is valid and points to an on-stack region.  Zeroing structures to avoid leaking kstack bits in padding will always be needed regardless of 'addr' being conflated.
